src/cargo/core/compiler/mod.rs

```diff
@@ -745,8 +745,10 @@ fn build_base_args<'a, 'cfg>(
     add_path_args(bcx, unit, cmd);
     add_color(bcx, cmd);
 
-    if bcx.build_config.json_messages() {
-        cmd.arg("--error-format").arg("json");
+    match bcx.build_config.message_format {
```

Comment:
When you rebase, I think it would be nice to pull this out into a separate function so that both the `rustc` and `rustdoc` calls could share the same code.